parser/src/main/kotlin/com/linkedin/dex/parser/JUnit4Extensions.kt

```diff
@@ -16,47 +16,75 @@ import com.linkedin.dex.spec.MethodIdItem
-        // For example, TestClass1 defines a test, EmptyClass2 extends TestClass1 and defines nothing, and then TestClass2
-        // extends EmptyClass2, TestClass2 should also list the tests defined in TestClass1
-        val classesWithAnnotations = dexFile.classDefs.filterNot(::isInterface)
-
-        classesWithAnnotations.map { classDef ->
-            val baseTests = dexFile.createTestMethods(classDef, dexFile.findMethodIds()).filter {
-                it.annotations.map {
-                    it.name
-                }.contains(testAnnotationName)
+    // Map for the second iteration to cache all found test methods including methods from superclass
+    val classAllTestMethods: MutableMap<String, Set<TestMethod>> = hashMapOf()
+
+    dexFiles
+            .forEach { dexFile ->
+
+                // We include classes that do not have annotations because there may be an intermediary class without tests
+                // For example, TestClass1 defines a test, EmptyClass2 extends TestClass1 and defines nothing, and then TestClass2
+                // extends EmptyClass2, TestClass2 should also list the tests defined in TestClass1
+                dexFile
+                        .classDefs
+                        .asSequence()
+                        .filterNot(ClassDefItem::isInterface)
+                        .forEach { classDef ->
+                            val testMethods = dexFile
+                                    .createTestMethods(classDef, dexFile.findMethodIds())
+                                    .filter { it.containsTestAnnotation }
+
+                            classTestMethods[dexFile.getClassName(classDef)] = ClassParsingResult(
+                                    dexFile = dexFile,
+                                    classDef = classDef,
+                                    className = dexFile.getClassName(classDef),
+                                    superClassName = dexFile.getSuperclassName(classDef),
+                                    testMethods = testMethods.toSet(),
+                                    isConcrete = classDef.isConcrete
+                            )
+                        }
             }
 
-            val superTests = createTestMethodsFromSuperMethods(dexFile, classDef, dexFile.formatClassName(classDef), getSuperTestMethods(classDef, classTestMethods, dexFile))
-            classTestMethods[dexFile.getClassName(classDef)] = ClassParsingResult(dexFile.getSuperclassName(classDef), baseTests union superTests, !(isAbstract(classDef) || isInterface(classDef)))
-        }
-    }
+    classTestMethods
+            .keys
+            .forEach { key ->
+                classTestMethods.computeIfPresent(key) { _, value ->
+                    val testMethods = createTestMethodsFromSuperMethods(value, classTestMethods, classAllTestMethods)
+                    value.copy(testMethods = value.testMethods union testMethods)
```

Comment:
I think it's better to keep classTestMethods immutable

Reply:
👍